Реализован функционал по ТЗ №8

OlegDyuzhin opened the pull request.

OlegDyuzhin committed:
Реализован функционал по ТЗ №8

avfyodorov commented on src/controllers/InMemoryTaskManager.java:
```diff
@@ -167,4 +173,42 @@ public void checkEpicStatus(int epicId) {
+    }
+
+    public Set<Task> getPrioritizedTask() {
+        return prioritizedTask;
```
Возвращать нужно список. По техническому заданию: 

.=============
Напишите новый метод getPrioritizedTasks, возвращающий список задач и подзадач в заданном порядке.
.==========
```java
	@Override
	public List<Task> getPrioritizedTasks() {
		return new ArrayList<>(prioritizedTasks);
	}
```

avfyodorov commented on src/controllers/InMemoryTaskManager.java:
```diff
@@ -167,4 +173,42 @@ public void checkEpicStatus(int epicId) {
+        epic.setEndTime(endTime);
+
+
+        if (startTime != null || endTime != null) epic.setDuration(Duration.between(startTime, endTime));
```
if (startTime != null || endTime != null) epic.setDuration(Duration.between(startTime, endTime));
    
Не совсем так. По техническому заданию: 
.===============
А вот с классом Epic нужно поработать дополнительно. Продолжительность эпика — сумма продолжительностей всех его подзадач.

avfyodorov commented on src/controllers/InMemoryTaskManager.java:
```diff
@@ -67,6 +69,7 @@ public void clearSubtasks() {
         subtasks.clear();
         for (Integer ids : epics.keySet()) {
             checkEpicStatus(ids);
+            checkEpicStartTimeAndDuration((Epic) getTaskById(ids));
```
```java
public void clearTasks() {
        tasks.clear();
    }
    @Override
    public void clearEpics() {
        epics.clear();
        subtasks.clear();
    }
    @Override
    public void clearSubtasks() {
        for......
```
.==================
Здесь и дальше. При удалении задач и подзадач нужно также не забыть удалить их из отсортированного по времени списка.

avfyodorov commented on src/controllers/InMemoryTaskManager.java:
```diff
@@ -138,12 +148,8 @@ public void removeTaskById(int id) {
     }
```
При удалении отдельной подзадачи или всех подзадач определенного Эпика также нужно удалить и из отсортированного по времени списка 

prioritizedTask.remove(

avfyodorov commented on src/controllers/InMemoryTaskManager.java:
```diff
@@ -86,26 +89,31 @@ public Task getTaskById(int id) {
             case SUBTASK -> subtasks.put(task.getId(), (Subtask) task);
             case EPIC -> epics.put(task.getId(), (Epic) task);
         }
+        if (task.getStartTime() != null) prioritizedTask.add(task);
```
if (task.getStartTime() != null) prioritizedTask.add(task);
    
Эпик добавлять не нужно, добавляются только задачи и подзадачи

OlegDyuzhin replied:
Поправил, но впринципе мало вероятно что будет подаваться эпик у которого уже будет задан startTime (такого конструктора нет). Но согласен это убережет от возможных дальнейших ошибок если данный конструктор появится =)

avfyodorov replied:
Да, но... лучше не давать такую возможность. Для надёжности. 🙂

avfyodorov reviewed (commented):
Добрый вечер, Олег! 

В целом хорошо, правда, есть ряд уточнений.

OlegDyuzhin committed:
Устранены замечания.
- переделан расчет duration Эпика
- добавлены удаления задач из prioritizedTask
- дописаны тесты для проверки методов по замечаниям

avfyodorov commented on src/controllers/InMemoryTaskManager.java:
```diff
@@ -167,4 +190,46 @@ public void checkEpicStatus(int epicId) {
+                .orElse(null);
+        epic.setEndTime(endTime);
+
+        Duration duration = getSubtaskEpic(epic).stream()
```
В принципе, всё это можно было сделать за один проход в одном цикле, без трёх новомодных стримов. 🙂

Для информации.

OlegDyuzhin replied:
Пометки замечены) буду ставить)

avfyodorov reviewed (commented):
Добрый день, Олег!

Всё в порядке,   замечаний нет.
Работа принята.

И ставьте, пожалуйста, пометки, чтобы было понятно, что комментарии замечены. )
